File: behavioral/mediator/mediator.java
public class MediatorPattern {
    public static class Cat {
        private String name;

        public Cat(String name) {
            this.name = name;
        }

        public void walk() {
            String result = String.format("%s is walking.", name);
            System.out.println(result);
        }

        public void sleep() {
            String result = String.format("%s is sleeping.", name);
            System.out.println(result);
        }
    }

    public static class Period {
        private boolean day = false;
        private Mediator mediator;

        public Period(Mediator mediator) {
            this.mediator = mediator;
        }

        public boolean getDay() {
            return day;
        }

        public void changeDay() {
            day = !day;
            String period = day ? "Day" : "Night";

            String result = String.format("%s has come.", period);
            System.out.println(result);

            mediator.dayChanged(day);
        }
    }

    public static class Weather {
        private boolean isRaining = false;
        private Mediator mediator;

        public Weather(Mediator mediator) {
            this.mediator = mediator;
        }

        public boolean getWeather() {
            return isRaining;
        }

        public void changeWeather() {
            isRaining = !isRaining;
            String verb = "is";

            if (!isRaining) {
                verb += " not";
            }

            String result = String.format("It %s raining.", verb);
            System.out.println(result);

            mediator.weatherChanged(isRaining);
        }
    }

    public static class Mediator {
        private Cat cat;
        private Period period;
        private Weather weather;

        public void setCat(Cat cat) {
            this.cat = cat;
        }

        public void setPeriod(Period period) {
            this.period = period;
        }

        public void setWeather(Weather weather) {
            this.weather = weather;
        }

        public void dayChanged(boolean day) {
            boolean raining = weather.getWeather();
            operateCat(day, raining);
        }

        public void weatherChanged(boolean raining) {
            boolean day = period.getDay();
            operateCat(day, raining);
        }

        private void operateCat(boolean day, boolean raining) {
            if (day && !raining) {
                cat.walk();
            } else {
                cat.sleep();
            }
        }
    }

    public static void main(String[] args) {
        Cat simba = new Cat("Simba");
        Mediator mediator = new Mediator();
        Period period = new Period(mediator);
        Weather weather = new Weather(mediator);

        mediator.setCat(simba);
        mediator.setPeriod(period);
        mediator.setWeather(weather);

        int iterations = 3;

        for (int i = 0; i < iterations; i++) {
            period.changeDay();

            int lastIteration = iterations - 1;
            if (i != lastIteration) {
                weather.changeWeather();
            }
        }
    }
}

File: behavioral/mediator/mediator.cs
using System;

class MediatorPattern
{
    class Cat
    {
        private string name { get; }

        public Cat(string name)
        {
            this.name = name;
        }

        public void Walk()
        {
            Console.WriteLine("{0} is walking.", name);
        }

        public void Sleep()
        {
            Console.WriteLine("{0} is sleeping.", name);
        }
    }

    class Period
    {
        private bool day { get; set; } = false;
        private Mediator mediator { get; }

        public Period(Mediator mediator)
        {
            this.mediator = mediator;
        }

        public bool GetDay()
        {
            return day;
        }

        public void ChangeDay()
        {
            day = !day;
            string period = day ? "Day" : "Night";

            Console.WriteLine("{0} has come.", period);

            mediator.DayChanged(day);
        }
    }

    class Weather
    {
        private bool isRaining { get; set; } = false;
        private Mediator mediator { get; }

        public Weather(Mediator mediator)
        {
            this.mediator = mediator;
        }

        public bool GetWeather()
        {
            return isRaining;
        }

        public void ChangeWeather()
        {
            isRaining = !isRaining;
            string verb = "is";

            if (!isRaining)
            {
                verb += " not";
            }

            Console.WriteLine("It {0} raining.", verb);

            mediator.WeatherChanged(isRaining);
        }
    }

    class Mediator
    {
        private Cat cat { get; set; }
        private Period period { get; set; }
        private Weather weather { get; set; }

        public void SetCat(Cat cat)
        {
            this.cat = cat;
        }

        public void SetPeriod(Period period)
        {
            this.period = period;
        }

        public void SetWeather(Weather weather)
        {
            this.weather = weather;
        }

        public void DayChanged(bool day)
        {
            bool raining = weather.GetWeather();
            OperateCat(day, raining);
        }

        public void WeatherChanged(bool raining)
        {
            bool day = period.GetDay();
            OperateCat(day, raining);
        }

        private void OperateCat(bool day, bool raining)
        {
            if (day && !raining)
            {
                cat.Walk();
            }
            else
            {
                cat.Sleep();
            }
        }
    }

    static void Main(string[] args)
    {
        Cat simba = new Cat("Simba");
        Mediator mediator = new Mediator();
        Period period = new Period(mediator);
        Weather weather = new Weather(mediator);

        mediator.SetCat(simba);
        mediator.SetPeriod(period);
        mediator.SetWeather(weather);

        int iterations = 3;

        for (int i = 0; i < iterations; i++)
        {
            period.ChangeDay();

            int lastIteration = iterations - 1;
            if (i != lastIteration)
            {
                weather.ChangeWeather();
            }
        }
    }
}
